File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.cpp
class Solution {
public:
    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        vector<int>dp(arr.size()+1,0);
        for(int i=1;i<=arr.size();i++)
        {
           int j=i;
            int a=k;
            int ma=arr[i-1];
            while(j>0 && a--)
            {
                ma=max(ma,arr[j-1]);
                dp[i]=max(dp[i],(i-j+1)*ma+dp[j-1]);
                j--;
            } 
        }
       
        return dp[arr.size()];
        
    }
};

File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.java
class Solution {
    public int maxSumAfterPartitioning(int[] arr, int k) {
        int [] dp=new int[arr.length+1];
        for(int i=1;i<=arr.length;i++){
            int x=arr[i-1];
            int cnt=1;
            for(int j=i;j>0 && j>i-k;j--){
                x=Math.max(arr[j-1],x);
                dp[i]=Math.max(dp[i],(x*cnt)+dp[j-1]);
                cnt++;
            }
            // System.out.print(dp[i]+" ");
        }
        return dp[arr.length];
        
        
    }
}